optimum/habana/transformers/models/falcon/modeling_falcon.py

```diff
@@ -373,17 +391,28 @@ def pre_attn_forward(
                         attn_output = self.sdpa(
                             query_layer, key_layer, value_layer, attention_mask, 0.0, is_causal=False
                         )
-                    else:
-                        with sdp_kernel(enable_recompute=False) if SDPContext else contextlib.nullcontext():
-                            attn_output = FusedSDPA.apply(
-                                query_layer,
-                                key_layer,
-                                value_layer,
-                                attention_mask,
-                                0.0,
-                                # The query_length > 1 is necessary to match with AttentionMaskConverter.to_causal_4d that does not create a causal mask in case query_length == 1.
-                                self.is_causal and attention_mask is None and query_length > 1,
-                            )
+                    else:  # use_flash_attention
+                        if query_length == 1:
+                            # next token
+                            use_recompute = True if os.getenv("QUANT_CONFIG", "") else False
+                            with sdp_kernel(enable_recompute=use_recompute):
+                                attn_output = self.fused_scaled_dot_product_attention(
+                                    query_layer, key_layer, value_layer, attention_mask, 0.0, False, None
+                                )
+                        else:
+                            # first token
+                            if flash_attention_causal_mask:
+                                # causal masking on first token requires inputs to be of the same lenght
+                                with sdp_kernel(enable_recompute=flash_attention_recompute):
+                                    attn_output = self.fused_scaled_dot_product_attention(
+                                        query_layer, key_layer, value_layer, attention_mask, None, 0.0, True, None
+                                    )
+                            else:
+                                with sdp_kernel(enable_recompute=flash_attention_recompute):
+                                    attn_output = self.fused_scaled_dot_product_attention(
+                                        query_layer, key_layer, value_layer, attention_mask, 0.0, False, None
+                                    )
+
```

Comment:
```
enable_recompute = (os.getenv("QUANT_CONFIG", "") != "") if query_length == 1 else flash_attention_recompute

bool = query_length == 1 or not flash_attention_causal_mask
dropout_p = 0.0 if bool else None
is_casual = False if bool else 0.0 # that looks strange.. ??
scale = ...


with sdp_kernel(enable_recompute=use_recompute):
      attn_output = self.fused_scaled_dot_product_attention(
          query_layer, key_layer, value_layer, attention_mask, dropout_p, is_casual, scale
      )
```


refactor to reduce number of lines



these are the inputs for the forward:
(self, query, key, value, attn_mask, dropout_p, is_casual, scale)

but this has more inputs:
 attn_output = self.fused_scaled_dot_product_attention(
                                        query_layer, key_layer, value_layer, attention_mask, **None, 0.0, True, None**
                                    )

Reply:
```
enable_recompute = (os.getenv("QUANT_CONFIG", "") != "") if query_length == 1 else flash_attention_recompute
is_causal = query_length != 1 and flash_attention_causal_mask
attn_mask = None if is_causal else attention_mask
with sdp_kernel(enable_recompute=enable_recompute):
      attn_output = self.fused_scaled_dot_product_attention(query_layer, key_layer, value_layer, attn_mask, 0.0, is_causal, None)
```